presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryUtils.java

```diff
@@ -73,6 +79,71 @@ public static GeometryTypeName valueOf(String type)
         }
     }
 
+    /**
+     * Copy of com.esri.core.geometry.Interop.translateFromAVNaN
+     *
+     * deserializeEnvelope needs to recognize custom NAN values generated by
+     * ESRI's serialization of empty geometries.
+     */
+    private static double translateFromAVNaN(double n)
+    {
+        return n < -1.0E38D ? (0.0D / 0.0) : n;
+    }
+
+    private static boolean isEsriNaN(double d)
+    {
+        return Double.isNaN(d) || Double.isNaN(translateFromAVNaN(d));
+    }
+
+    @Nullable
+    public static Envelope deserializeEnvelope(Slice shape)
+    {
+        if (shape == null) {
+            return null;
+        }
+        BasicSliceInput input = shape.getInput();
+
+        input.readInt(); // spatial reference ID
+
+        Envelope overallEnvelope = null;
+        while (input.available() > 0) {
+            int length = input.readInt();
+            ByteBuffer buffer = input.readSlice(length).toByteBuffer().order(LITTLE_ENDIAN);
+            int type = buffer.getInt();
+            Envelope envelope = null;
+            if (type == POINT_TYPE) {    // point
+                double x = buffer.getDouble();
+                double y = buffer.getDouble();
+                if (!isEsriNaN(x)) {
+                    Verify.verify(!isEsriNaN(y));
+                    envelope = new Envelope(x, y, x, y);
+                }
+            }
+            else {
+                double xMin = buffer.getDouble();
+                double yMin = buffer.getDouble();
+                double xMax = buffer.getDouble();
+                double yMax = buffer.getDouble();
+                if (!isEsriNaN(xMin)) {
+                    Verify.verify(!isEsriNaN(xMax));
```

Comment:
I wonder if it might have some perf impact...

Reply:
@nezihyigitbasi , @martint , do you know whether we should use Verify.verify or Preconditions.checkState ?

Reply:
If you are asking in general when to use which one, please see [this](https://github.com/google/guava/wiki/ConditionalFailuresExplained). There is also some good info in the `Verify` [javadoc](https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/base/Verify.html).

Reply:
@nezihyigitbasi , thank you for the pointers.